### PR TITLE
trace exec: add upper_layer field

### PR DESCRIPTION
**Pull request by alban**
# trace exec: add upper_layer field

The upper_layer field says whether the program file is on the upper layer of overlayfs, i.e. it was modified in the container.

## How to use

```
$ go run -exec 'sudo -E' ./cmd/ig/... trace exec
INFO[0000] Experimental features enabled                
RUNTIME.CONTAINERNAME PID    PPID   COMM  RET ARGS                                                       UPPERLAYER
reverent_austin       137509 137486 sh    0   /usr/bin/sh -c cp /bin/echo /bin/echo2 ; /bin/echo lower … 0
reverent_austin       137532 137509 cp    0   /usr/bin/cp /bin/echo /bin/echo2                           0
reverent_austin       137533 137509 echo  0   /bin/echo lower                                            0
reverent_austin       137534 137509 echo2 0   /bin/echo2 upper                                           1
```

```
$ docker run -ti --rm ubuntu sh -c 'cp /bin/echo /bin/echo2 ; /bin/echo lower ; /bin/echo2 upper'
lower
upper
```

As you can see in the output, only `/bin/echo2` is reported as being on the overlay upper layer.

## Testing done

See above.

## TODOs

- [x] Update documentation
- [x] Update tests
- [x] ~Should I use a bitfield?~ Use bools
- [x] ~How should the output look like for the user?~ Show booleans (true/false)

## Limitations

The reporting is incorrect if the exec failed (e.g. due to lack of +x permission). Example:

```
$ docker run -ti --rm ubuntu sh -c 'cat /bin/echo > /bin/echo2 ; /bin/echo lower ; /bin/echo2 upper'
lower
sh: 1: /bin/echo2: Permission denied
```

```
$ go run -exec 'sudo -E' ./cmd/ig/... trace exec -o columns=comm,ret,args,upperlayer
INFO[0000] Experimental features enabled                
COMM             RET ARGS                                                   UPPERLAYER
sh               0   /usr/bin/sh -c cat /bin/echo > /bin/echo2 ; /bin/echo… 0         
cat              0   /usr/bin/cat /bin/echo                                 0         
echo             0   /bin/echo lower                                        0         
sh               -13 /bin/echo2 upper                                       0
```

## Misc

Fixes https://github.com/inspektor-gadget/inspektor-gadget/issues/2345

See also #1913

cc @galofir-ms

### CODE REVIEW

File: pkg/gadgets/trace/exec/tracer/bpf/execsnoop.bpf.c [hunk 1]
```diff
@@ -139,6 +142,40 @@ int ig_execve_e(struct trace_event_raw_sys_enter *ctx)
...
+	}
+
+	struct _ovl_inode *oi = container_of(inode, struct _ovl_inode, vfs_inode);
+	if (!oi) {
+        return false;
+	}
+
+	struct dentry *upperdentry;
+
+	bpf_probe_read_kernel(&upperdentry, sizeof(struct dentry *), &oi->__upperdentry);
```
[the remark this answers is not in the document]
Comment (alban) [reply pkg/gadgets/trace/exec/tracer/bpf/execsnoop.bpf.c:175]: I am not sure to understand the suggestion. Here it is the size of a pointer (`void *`). What should I write?

**Comment by alban** [comment]
> The reporting is incorrect if the exec failed (e.g. due to lack of +x permission).

My plan is to document this limitation and work on it in a future PR.

**Comment by alban** [comment]
The CI fails because:
- README: temporarily dead links
- cri-o test failure on TestTraceNetwork:
```
2024-01-17T11:13:56.2788453Z         time="2024-01-17T11:13:45Z" level=error msg="namespace enricher: failed to get mnt namespace on container 34ce46b7bda2cb050eba4573a68ddcf893f6e76ab5c2acfea8fed797eced2d73: stat /proc/7347/ns/mnt: no such file or directory"
2024-01-17T11:13:56.2790989Z         time="2024-01-17T11:13:45Z" level=warning msg="WithTracerCollection: failed to open mntns reference for container 34ce46b7bda2cb050eba4573a68ddcf893f6e76ab5c2acfea8fed797eced2d73: no such file or directory"
2024-01-17T11:13:56.2793665Z         time="2024-01-17T11:13:45Z" level=error msg="cgroup enricher: failed to get cgroup paths on container b500d81aeadc4796941c5162c46087af882309aa2544463fa8a4cde8539f7930: parsing cgroup: open /proc/4377/cgroup: no such file or directory"
2024-01-17T11:13:56.2796070Z         time="2024-01-17T11:13:45Z" level=error msg="namespace enricher: failed to get mnt namespace on container b500d81aeadc4796941c5162c46087af882309aa2544463fa8a4cde8539f7930: stat /proc/4377/ns/mnt: no such file or directory"
2024-01-17T11:13:56.2798323Z         time="2024-01-17T11:13:45Z" level=warning msg="WithTracerCollection: failed to open mntns reference for container b500d81aeadc4796941c5162c46087af882309aa2544463fa8a4cde8539f7930: no such file or directory"
2024-01-17T11:13:56.2800277Z         time="2024-01-17T11:13:46Z" level=warning msg="start tracing container \"test-pod\": getting network namespace of pid 8048: stat /proc/8048/ns/net: no such file or directory"
2024-01-17T11:13:56.2801525Z         time="2024-01-17T11:13:46Z" level=warning msg="stop tracing container \"test-pod\": pid 8048 is not attached"
...
2024-01-17T11:13:56.3781612Z     helpers.go:145: output doesn't contain the expected entry
2024-01-17T11:13:56.3782065Z         captured:
...
2024-01-17T11:13:56.4611587Z         expected:
2024-01-17T11:13:56.4613753Z         {"runtime":{"runtimeName":"cri-o","containerName":"test-pod","containerImageName":"docker.io/library/busybox:latest"},"k8s":{"namespace":"test-trace-network-2779704655061405693","podName":"test-pod","containerName":"test-pod"},"type":"normal","comm":"wget","uid":0,"gid":0,"pktType":"OUTGOING","proto":"TCP","port":80,"dst":{"addr":"10.244.0.12","version":4}}
```

Both are unrelated to this PR.

File: pkg/gadgets/trace/exec/tracer/bpf/execsnoop.bpf.c [hunk 2]
```diff
@@ -139,6 +142,43 @@ int ig_execve_e(struct trace_event_raw_sys_enter *ctx)
...
+
+	struct _ovl_inode *oi =
+		container_of(inode, struct _ovl_inode, vfs_inode);
+	if (!oi) {
+		return false;
+	}
+
+	struct dentry *upperdentry;
+
+	bpf_probe_read_kernel(&upperdentry, sizeof upperdentry,
```
Comment (mauriciovasquezbernal) [inline pkg/gadgets/trace/exec/tracer/bpf/execsnoop.bpf.c:177]: Already commented in another PR some time ago. It's not mandatory but I'd rather use `sizeof(upperdentry)` to keep the code aligned.

**Comment by mauriciovasquezbernal** [comment]
@alban it seems it broke the CI on AKS and ARO:

## AKS

upperlayer for `sh` event is true, but it should be false:

captured:
```json
{
        "runtime": {
                "containerImageName": "docker.io/library/busybox:latest"
        },
        "k8s": {
                "namespace": "test-trace-exec-7345717025894922932",
                "podName": "test-pod",
                "containerName": "test-pod"
        },
        "type": "normal",
        "comm": "sh",
        "args": [
                "/bin/sh",
                "-c",
                "cp /bin/date /date ; setuidgid 1000:1111 sh -c 'while true; do /date ; /bin/sleep 0.1; done'"
        ],
        "uid": 0,
        "gid": 0,
        "upperlayer": true,
        "loginuid": 0,
        "sessionid": 0,
        "cwd": "/"
}
```
expected:
```json
{
        "runtime": {
                "containerImageName": "docker.io/library/busybox:latest"
        },
        "k8s": {
                "namespace": "test-trace-exec-7345717025894922932",
                "podName": "test-pod",
                "containerName": "test-pod"
        },
        "type": "normal",
        "comm": "sh",
        "args": [
                "/bin/sh",
                "-c",
                "cp /bin/date /date ; setuidgid 1000:1111 sh -c 'while true; do /date ; /bin/sleep 0.1; done'"
        ],
        "uid": 0,
        "gid": 0,
        "upperlayer": false,
        "loginuid": 0,
        "sessionid": 0,
        "cwd": "/"
}
```

https://github.com/inspektor-gadget/inspektor-gadget/actions/runs/7572632543/attempts/1#summary-20624245156
https://github.com/inspektor-gadget/inspektor-gadget/actions/runs/7572632543/attempts/1#summary-20624245512

## ARO

Event for `/date` has upperlayer=false, but it should be true.

captured:
```json
{
        "runtime": {
                "containerImageName": "docker.io/library/busybox:latest"
        },
        "k8s": {
                "namespace": "test-trace-exec-5865986315796987308",
                "podName": "test-pod",
                "containerName": "test-pod"
        },
        "type": "normal",
        "comm": "date",
        "args": [
                "/date"
        ],
        "uid": 1000,
        "gid": 1111,
        "upperlayer": false,
        "loginuid": 0,
        "sessionid": 0,
        "cwd": "/"
}
```

expected:
```json
{
        "runtime": {
                "containerImageName": "docker.io/library/busybox:latest"
        },
        "k8s": {
                "namespace": "test-trace-exec-5865986315796987308",
                "podName": "test-pod",
                "containerName": "test-pod"
        },
        "type": "normal",
        "comm": "date",
        "args": [
                "/date"
        ],
        "uid": 1000,
        "gid": 1111,
        "upperlayer": true,
        "loginuid": 0,
        "sessionid": 0,
        "cwd": "/"
}
```

https://github.com/inspektor-gadget/inspektor-gadget/actions/runs/7572632543/attempts/1#summary-20624245813

**Comment by alban** [comment]
I tested manually on AKS and it worked for me:
```
$ ./kubectl-gadget trace exec -o columns=k8s.node,k8s.container,comm,ret,args,upperlayer
INFO[0000] Experimental features enabled                
K8S.NODE                   K8S.CONTAINER              COMM           RET ARGS                               UPPERLAYER
aks-userpool-…5-vmss000000 shell01                    date           0   /usr/bin/date                      false     
aks-userpool-…5-vmss000000 shell01                    date           0   /date                              true
```

```
$ kubectl run -ti --rm --image=ubuntu --restart=Never shell01 -- sh
# bash
root@shell01:/# cp /usr/bin/date /date
root@shell01:/# date ; /date
root@shell01:/# uname -a
Linux shell01 5.15.0-1053-azure #61-Ubuntu SMP Tue Nov 21 14:16:01 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
```

I will test with a azure nodepool instead of a ubuntu nodepool.

**Comment by alban** [comment]
Kernel configuration:
```
config RANDSTRUCT
        def_bool !RANDSTRUCT_NONE
config GCC_PLUGIN_RANDSTRUCT
        def_bool GCC_PLUGINS && RANDSTRUCT
        help
          Use GCC plugin to randomize structure layout.
```

On my laptop:
```
$ cat /boot/config-`uname -r`|grep RANDSTRUCT
CONFIG_RANDSTRUCT_NONE=y
$ cat /boot/config-`uname -r`|grep CONFIG_GCC_PLUGINS  
# CONFIG_GCC_PLUGINS is not set
```
On the Azure node:
```
# cat /boot/config-`uname -r`|grep RANDSTRUCT
# CONFIG_GCC_PLUGIN_RANDSTRUCT is not set
# cat /boot/config-`uname -r`|grep CONFIG_GCC_PLUGINS  
CONFIG_GCC_PLUGINS=y
```

In our current code, we rely on fields relative positions:
https://github.com/inspektor-gadget/inspektor-gadget/blob/8c433c4b29399cabd6ac233933c602f76c3c7c5f/pkg/gadgets/trace/exec/tracer/bpf/execsnoop.bpf.c#L150

But we cannot rely on fields relative positions in a struct when fields positions are randomized. It might or might not work, depending on how the random order was selected at boot time.

**Comment by alban** [comment]
`struct ovl_inode` does not have `__randomize_layout` (and it does not consist of [entirely function pointers](https://github.com/torvalds/linux/blob/v6.2/security/Kconfig.hardening#L290-L293) causing its layout to be randomized) , so relying on relative positions should work fine. But `struct inode` does (see [include/linux/fs.h#L733](https://github.com/torvalds/linux/blob/v5.15/include/linux/fs.h#L733)). I wonder if `__randomize_layout` causes a change of size (`sizeof`) or of packing, which would cause the problem.

According to [this lkml email](https://lore.kernel.org/bpf/202003301016.D0E239A0@keescook/T/#m27f370d6453d86e11cce7de89c5f6aae7ec64b90), `GCC_PLUGIN_RANDSTRUCT` generates invalid BTF. I tested with `bpftool btf dump file /tmp/vmlinux-5.15.138.1-4.cm2 format c` that `/sys/kernel/btf/vmlinux` (copied from the azure node) does not know the correct randomization for `struct inode`.

**Pull request merged**